Changes to fix issues of missing extensions and others

chandrashekar-s opened the pull request:
## Description of what I changed

Fixes #1027 
The below mentioned changes have been made as part of this PR
- Handled the error properly when the StructureDefinitions for extensions are missing.
- The DoubleConverter for decimal data types has been fixed, which was failing for Leaf Extensions containing `Decimal` data types.
- The avro schema name for the `Reference` types have been changed to contain a full path name from root, so that it is unique for each path.

## E2E test

- Tested the changes by referencing an extension whose StructureDefinition was not available and ensure proper error message is thrown instead of NPE.
- Ingested an `QuestionnaireResponse` fhir resource and ran a `Full Run` and checked if the `QuestionnaireResponse` is converted to `Avro` record properly

TESTED:

- Added a unit test case

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

bashir2 commented on bunsen/bunsen-core-r4/src/main/java/com/cerner/bunsen/definitions/r4/R4StructureDefinitions.java:
```diff
@@ -38,8 +39,11 @@ public FhirConversionSupport conversionSupport() {
-        (StructureDefinition) context.getValidationSupport().fetchStructureDefinition(resourceUrl));
+    IBaseResource baseResource =
+        context.getValidationSupport().fetchStructureDefinition(resourceUrl);
+    return baseResource == null
```
I see that in many cases that call this method, you are catching the `null` case and throwing a `RuntimeException` for it; is there a reason not to throw such exceptions here? I mean is there a case where we can do anything else in the `null` case?

bashir2 replied:
What is the contract of `getStructureDefinition`; I mean is `null` an expected/possible return value? Asking because now after looking a little bit closer, I am more concerned if `getStructureDefinition` can return `null`. For example [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java#L316) we don't handle `null` and potentially have an NPE.

So here is my suggestion: Either (i) establish that `null` is a valid return value of `getStructureDefinition` and then handle `null` everywhere this is called; OR (ii) add @NotNull as part of its return contract (in the base interface) and in both R4 and Stu3 implementations, check that we do not return `null` (e.g., throw an exception).

bashir2 replied:
But I think the question still remains, is returning `null` an expected return value of `getStructureDefinition`? To clarify further, the example you mentioned earlier from [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java#L621), is not really the return value of `getStructureDefinition` but `getDefinition`. In other words, if `getDefinition` return `null` [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java#L338) it seems to me that it is assumed the type is `BackboneElement` or `Element`. Now if we change the contract of `getStructureDefinition` we are breaking that assumption.

So to summarize, I suggest that we never return `null` from `getStructureDefinition` and clean up all the exceptions that are added (move them to a single place in `getStructureDefinition`), unless if there is a legitimate case for which I appreciate if a unit-test is added demonstrating that `getStructureDefinition` returning `null` is a legitimate case.

chandrashekar-s replied:
Ok, makes sense. In the latest commit, I have added the **@Nonnull** annotation and a `@throws IllegalArgumentException` to the method contract in both the interface and implementations. Though the `@throws` declaration was not mandatory (as its not a checked exception), I have added for the sake of clarity.

bashir2 commented on pipelines/common/src/test/resources/questionnaire_response.json:
```diff
@@ -0,0 +1,83 @@
+  "id" : "f201",
+  "text" : {
+    "status" : "generated",
+    "div" : "<div xmlns=\"http://www.w3.org/1999/xhtml\"><table border=\"1\" cellpadding=\"0\" cellspacing=\"0\" style=\"border: 1px #F0F0F0 solid; font-size: 11px; font-family: verdana; vertical-align: top;\"><tr style=\"border: 2px #F0F0F0 solid; font-size: 11px; font-family: verdana; vertical-align: top\"><th style=\"vertical-align: top; text-align : left; background-color: white; border: 1px  ...
```
Where have you got this from? Also if not needed, remove the `div` field to make this smaller.

chandrashekar-s replied:
I could not remember where I got this from, in the latest commit I have replaced this with one of the [examples](https://hl7.org/fhir/us/core/STU6.1/QuestionnaireResponse-glascow-coma-score.json.html) from HL7 portal. Also, removed the `div` field.

bashir2 commented on bunsen/bunsen-avro/src/test/resources/r4-custom-schemas/bunsen-test-patient-schema.json:
```diff
@@ -562,6 +561,228 @@
                         "default": null
                       },
+                      {
+                        "name": "identifier",
```
As discussed offline, can you please do some minimal memory profiling between the old and new schema and see how much impact this larger schema has on the memory footprint?

chandrashekar-s replied:
I did some basic memory profiling between the old and new schema and it looks like there is an increase of roughly by 20% in the size of parquet files in the new schema. This is because of the recent changes where in by default now we capture a nested field size of 2 for repeated fields (earlier it was 1 due to a bug which is fixed in this PR). This reminds us to relook at the `recursiveDepth` parameter and do the right recursive capture. I have created this [issue](https://github.com/google/fhir-data-pipes/issues/1062) to relook into this.

On the other hand, due to the changes in this [PR](https://github.com/google/fhir-data-pipes/pull/1047) to flush the parquet files in the finishBundle method, it has created some regression on the number of parquet files generation for `DirectRunner`. Earlier (before this [PR](https://github.com/google/fhir-data-pipes/pull/1047) changes), the number of parquet files created for each resource type is equal to the `parallelism` parameter being passed in the `FhirEtlOptions`, whereas after the change the number of files created is equal to the number of FHIR resources i.e. for a total of 12K observations, it created 12K parquet files. This was causing an `OutOfMemoryError` while checking number of the records in the parquet files using `parquet-tools` util jar in the e2e tests.

I have created this [issue](https://github.com/google/fhir-data-pipes/issues/1063) for the above regression.

chandrashekar-s replied:
I have increased the heap size in the current PR to check the number of FHIR resources in parquet files. This has to be reverted back once this [issue](https://github.com/google/fhir-data-pipes/issues/1063) is fixed.

Interestingly, the e2e-tests was not failing in the PR where the changes were made to fix the parquet file generation for DataFlowRunner. This was because the default max heap size (8gb) was sufficient enough to check the count, but with the recent increase in schema size i.e. ~20%, this step is failing in the current PR.

bashir2 replied:
Thanks for checking and the explanation.
For future reference (following our off-PR discussion): I think the 20% increase is for the small dataset we use for e2e test. For larger/more realistic datasets, this ratio should be much smaller, correct?

chandrashekar-s replied:
Yes @bashir2, for larger dataset the difference seems to be negligible. I verified this for the 8K Patients dataset and here is the data size of the parquet files before and after the changes.

Before Change
--------------
Patients - 2.7MB
Encounter - 86MB
Observation - 123MB

After Change
-------------
Patients - 3MB
Encounter - 86MB
Observation - 124MB

So, I am assuming that for smaller dataset due to the difference in schema size, the difference is visible, but as the data grows this schema difference seems to be negligible (unless even the data is captured with so many levels)

bashir2 reviewed (approved):
Thanks @chandrashekar-s for fixing these issues. All of my comments are either minor or questions, so please feel free to merge this after you address them and the e2e issue is fixed.

chandrashekar-s commented on bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java:
```diff
@@ -545,12 +553,24 @@ private <T> T transform(
       List<String> referenceProfiles = parentElement.getReferenceTargetProfiles();
```
Looks like special handling for `Reference` types is still needed as we add all the `references` and the `children` to the converter as done over [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java#L712) and [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java#L735). That said, this has been the code since beginning. The fix that was done to this branch was over [here](https://github.com/google/fhir-data-pipes/pull/634/files#diff-3074f63a61ae44ee27049eccc58e6ae935ffe5a35e159de6c533cb751a6b1fe7), where a unique name was supposed to be generated but it was ending up with `Identifier.assignerReference` irrespective of the stack path. This has been fixed in the current PR, where the name depends on the path from the root in the format `<PathFromRoot>+<ElementName>Reference`  e.g. `PatientReferenceIdentifierAssignerReference` will be name for `Patient.generalPractitioner.identifier.assigner` and `Patient.managingOrganization.identifier.assigner`. However the names for `Patient.generalPractitioner` and `Patient.managingOrganization` are different i.e. `PatientGeneralPractitionerReference` and `PatientManagingOrganizationReference` respectively. This makes sense because the `Patient.generalPractitioner` references multiple reference types and should be different from `Patient.managingOrganization`.

bashir2 replied:
I see, thanks for checking and the clarification. Yeah, we need to investigate more that increase in schema size and take action if necessary (after this PR).

chandrashekar-s replied:
sure

chandrashekar-s commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java:
```diff
@@ -102,7 +102,7 @@ public Schema getDataType() {
   // Note with this approach we are loosing some precision!
   private static final HapiConverter<Schema> DOUBLE_CONVERTER =
-      new PrimitiveConverter<Schema>("Double") {
+      new PrimitiveConverter<Schema>("Decimal") {
```
yes, for primitive types it is used in `LeafExtensionConverter` and also during merging of converters in which case they are used for just validation. Updated the documentation in `HapiConverter` as well.

chandrashekar-s commented:
Hi @bashir2, thanks for the review. I have updated my comments and I have also created these 2 issues ([Relook at the recursive depth](https://github.com/google/fhir-data-pipes/issues/1062) and [Fix the number of parquet files for DirectRunner](https://github.com/google/fhir-data-pipes/issues/1063)) which can be resolved in different PRs.

bashir2 reviewed (approved):
Thanks @chandrashekar-s for the detailed responses. My only main remaining issue is the `null` return value. Otherwise, please feel free to merge this.

chandrashekar-s commented on e2e-tests/controller-spark/controller_spark_sql_validation.sh:
```diff
@@ -356,7 +356,9 @@ clear
-sleep 10
+# Provide enough buffer time before triggering the incremental run so that the previous full run
+# completes fully (including creation of hive tables)
+sleep 60
```
Sometimes it is observed that the incremental run does not gets triggered because the previous full run is still in progress (this is because we only wait for generation of parquet files and not the creation of hive tables). One such example is this [build failure](https://ff3374279beb0c9b34708ea31d07a39db7f25084ae26d62a05a5c89-apidata.googleusercontent.com/download/storage/v1/b/cloud-build-gh-logs/o/log-b5056f60-0f4a-4a0a-a4d9-94ad1018d88f.txt?jk=ASOpP9gaDsevegR6H21KWVsBiPZw_NnweQRuvXmKms43rJdcV77acJO-G5Y_qoyGjSmDEfhKRpBcja6uzsSeypgYQFdL92zTr6NnOQf8py7ijlgQ1e8Vu-_QRADE0tmWHXx-8bbmebRsQANXHBtLjhYuAPue3AyjNZHUN0wdHsKG41r-Mq7YxvbyLDTQxdgsPNTvcic8AasH0M3Lc9JWDHR5OuVO707YUHL7PfejpllYDx8OjNL2PaZdTXfABZvxOUWturqhpsXHGBQ7hNLTQftSfZy6Yrxx4UyfJfI3OU68g7DYQm1u6UM_wz1wJPYx7M1xoctBpb3mocnPPRxl0MFQH9oCVtPc7z1KrLSvVyJ8sAAZ5MFKsw1cpyWP0yllq7ReTzpVkI9qckIM57Uqc3ZOzOWkBJPOW4op2wravNgxX21V_eInfq80e0-YCrv0M6iv4AlZr1Yc9UovXEzo9b1JNfQP3msyzR7HrFUBf1wJ48DMV-g90sVPXx70m1EfdR04oBsSH5f3hZ745Z_BlbfP-Hp7o8fZfjGd_TZVsv-HhI1-MDjjwwrudW12BvOtOsRTLlCB1ayED6BXQmUsVfrvbqa64Rw8CnDI38OEpgtJ87k8wJFOpWLtK5uJ4taztmjpMYasg80LpYtwnaoL8SKoc9bgUc8VKX0eFKbWFgw0NgTTTygK1Gc3VAuQfALz6ktml0h2pL3X_wH7BwjazJMFLrW5QZsLjOXtCv4LexrQJoxQGMcwlBfHUhQYwqFU3JDicWUbIe4lottjiQ3b40yGfhUNVeaWxD0uCxq2Qp00R_e8G2t7z_-BtJCFDRgi3kXqSAxCrR3y6jDJRprmUylc9kecnLUONravRUjJXjZWhkyju0k9k0049MdWR4QK86dtQaPfL9DKpGdfsY7rVqdBj9Jwwyah_mgfpysQmJ5PvevpbZwXttjYybhUShbfRfkHd5MarjneFKtCpVdG0iwVS7UaMj8bhScIt5NR2MByjnkZW18jlebvaHMarCi35fAqKScDGyebq1I8QRksfSi21_uTWZ42pDDkrhY8lyT0LY2yDz78-ZRhzHwa7D55frUT_0D9l4RgSk_h06lAWE8DX6d0b2h0s-TomRU9ZK1hq-Ys6mYJCeMSF5Kbgd2lH3bS-7jX1pjMTX9LaF1NPsQyIxjxiqO4nwn3o5iuCd7JA70FhKt9ouG_3lNmxEF2i0Vovke-9vITnnPicRKmkmMpzNs2zyHVBIe1Qj4&isca=1).

chandrashekar-s commented:
@bashir2, I have addressed the review comments and will be merging the PR. Please let me know if you see any discrepancies and I will take them up in a separate PR.
